stdlib/fun.mli

```diff
@@ -31,3 +33,30 @@ val flip : ('a -> 'b -> 'c) -> ('b -> 'a -> 'c)
 val negate : ('a -> bool) -> ('a -> bool)
 (** [negate p] is the negation of the predicate function [p]. For any
     argument [x], [(negate p) x] is [not (p x)]. *)
+
+(** {1 Exception handling} *)
+
+val protect : finally:(unit -> unit) -> (unit -> 'a) -> 'a
+(** [protect ~finally work] invokes [work ()] and then [finally ()]
+    before [work ()] returns with its value or an exception. In the
+    latter case the exception is re-raised after [finally ()]. If
+    [finally ()] raises an exception, then the exception
+    {!Finally_raised} is raised instead.
+
+    [protect] can be used to enforce local invariants whether the
+    function returns normally or raises an exception. However, it does
+    not protect against unexpected exceptions raised inside [finally
+    ()] such as {!Stdlib.Out_of_memory}, {!Stdlib.Stack_overflow}, or
+    asynchronous exceptions raised by signal handlers
+    (e.g. {!Sys.Break}). *)
+
+exception Finally_raised of exn
+(** Raised by [protect ~finally work] when [finally] raises an
+    exception. The argument is the exception that was raised.
+
+    [Finally_raised] can contain an unexpected exception raised in
+    [finally] and/or shadow one raised in [work], such as
+    {!Stdlib.Invalid_argument}, {!Stdlib.Out_of_memory},
+    {!Stdlib.Stack_overflow}, or any asynchronous exception raised by
+    finalisers or signal handlers (e.g. {!Sys.Break}); for this reason
+    it should be considered as serious as those. *)
```

Comment:
If you don't want to spell out what should be done (not catch it) then I'd rather remove the last sentence, it brings nothing, even more people will think that if it's serious it should be treated.

Reply:
That's a possible solution. If I had to spell out exactly what one can do, it would be more complicated than "not catch it", and I am not sure everybody would agree.

Reply:
> it would be more complicated than "not catch it", and I am not sure everybody would agree.

There's no need to play cat and mouse... simply tell what you'd want to write (or actually write it) and you'll see what people respond. Otherwise just remove the sentence, it's needlessly distracting.